sklearn/tree/_tree.pyx

```diff
@@ -1230,3 +1228,399 @@ cdef class Tree:
             if not (0.999 < total_weight < 1.001):
                 raise ValueError("Total weight should be 1.0 but was %.9f" %
                                  total_weight)
+
+
+# =============================================================================
+# Build Pruned Tree
+# =============================================================================
+
+
+cdef class _CCPPruneController:
+    """Base class used by build_pruned_tree_ccp to control pruning
+    """
+    cdef bint stop_pruning(self, DOUBLE_t effective_alpha) nogil:
+        """Return 1 to stop pruning and 0 to continue pruning"""
+        return 0
+
+    cdef int save_metrics(self, DOUBLE_t effective_alpha,
+                          DOUBLE_t subtree_impurities) nogil:
+        """Save metrics when pruning"""
+        pass
+
+cdef class _AlphaPruner(_CCPPruneController):
+    """Use alpha to control when to stop pruning"""
+    cdef DOUBLE_t ccp_alpha
+
+    def __cinit__(self, DOUBLE_t ccp_alpha):
+        self.ccp_alpha = ccp_alpha
+
+    cdef bint stop_pruning(self, DOUBLE_t effective_alpha) nogil:
+        # The subtree on the previous iteration has the greatest ccp_alpha
+        # less than or equal to self.ccp_alpha
+        return self.ccp_alpha < effective_alpha
+
+
+cdef class _PathFinder(_CCPPruneController):
+    """Record metrics used to return the cost complexity path"""
+    cdef DOUBLE_t[:] ccp_alphas
+    cdef DOUBLE_t[:] impurities
+    cdef UINT32_t count
+
+    def __cinit__(self,  int node_count):
+        self.ccp_alphas = np.zeros(shape=(node_count), dtype=np.float64)
+        self.impurities = np.zeros(shape=(node_count), dtype=np.float64)
+        self.count = 0
+
+    cdef int save_metrics(self,
+                          DOUBLE_t effective_alpha,
+                          DOUBLE_t subtree_impurities) nogil:
+        self.ccp_alphas[self.count] = effective_alpha
+        self.impurities[self.count] = subtree_impurities
+        self.count += 1
+
+
+cdef _cost_complexity_prune(
+    Tree orig_tree,
+    _CCPPruneController controller,
+    unsigned char[:] leaves_in_subtree,
```

Comment:
in the other function you mark outputs with a `# OUT`, which I find really nice, you could add them here.

Reply:
This does not really have an `# OUT`. The controller stores all the state the the caller needs. I'll add a comment about this.

Reply:
I refactored this code to remove `capacity` and placed it into the controller. With the controller handling this, the pruning path will not calculate the capacity (since it is not used when finding the path).